DOCS-9381-dw-filter-example-kt (#1695)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/dw-core-functions-filter.adoc

----

=== Example

> This example reads a JSON array that contains `name` and `lastName` values, and uses the `filter` function to return the values that are not `null`.

Reviewer: ```suggestion
This example reads a JSON array that contains objects with `user` and `error` keys, and uses the `filter` function to return the objects that have their key `error` set with the value `null`.
```
If the script is changed to the suggested one.

==== Source
[source,DataWeave, linenums]
----
%dw 2.0
output application/json
---
[
   {
      "example":{
         "name":"username one",
         "lastName":"lastname one"
      },
      "error":"That name doesn't exists"
   },
   {
      "example":{
         "name":"username two",
         "lastName":"lastname two"
      },
      "error":null
   }
]

> filter ((item, index) -> item.error != null)

Reviewer: ```suggestion
%dw 2.0
output application/json

var users = [
   {
      "user": {
         "name": "123",
         "lastName": "Smith"
      },
      "error": "That name doesn't exists"
   },
   {
      "user": {
         "name": "John",
         "lastName": "Johnson"
      },
      "error": null
   }
]
---
users filter ((item, index) -> item.error == null)
```

I think that calling filter from a variable or payload is a more common use case than having the data in the body of the script. And maybe filter out the one with the error (item.error == null).

----

==== Output
[source,JSON,linenums]
----
[
  {
    "example": {
      "name": "username one",
      "lastName": "lastname one"
    },
    "error": "That name doesn't exists"
  }

> ]

Reviewer: ```suggestion
[
  {
    "user": {
      "name": "John",
      "lastName": "Johnson"
    },
    "error": null
  }
]
```
If the script is changed to the one suggested.

Reviewer 2: Looks like that change didn't happen.

----

=== Example

> This example reads a JSON array that contains objects with `user` and `error` keys, and uses the `filter` function to return the objects that have their key `error` set with the value `null`.

Reviewer 2: "to return the objects that have their key `error` set with the value `null`"  -> "to return only the objects in which the value of the `error` key is `null`" or something similar.

==== Source
[source,DataWeave, linenums]
----
%dw 2.0
output application/json

var users = [
   {
      "user": {
         "name": "123",
         "lastName": "Smith"
      },
      "error": "That name doesn't exists"
   },
   {
      "user": {
         "name": "John",
         "lastName": "Johnson"
      },
      "error": null
   }
]
---
users filter ((item, index) -> item.error == null)
----

==== Output
[source,JSON,linenums]
----

> [

Reviewer 2: Need to fix the output. Should be:
```
[
  {
    "user": {
      "name": "John",
      "lastName": "Johnson"
    },
    "error": null
  }
]
```
